### PR TITLE
fix: `phpdoc_types` must not lowercase literal types

### LINKED ISSUE #7101: phpdoc_types: 'NULL' vs. 'null'
## Bug report

phpdoc_types will convert `@return 'NULL'|null|false|int` into `@return null|'null'|false|int`, but this is not the same. 'NULL' vs. 'null'  

### Runtime version

PHP CS Fixer 3.19.2 (210dd89) Triangle by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 7.4.33

### Used command

 php /srv/www/foo/vendor/php-cs-fixer.phar \
        fix \
          \
        --diff \
        --ansi \
        -vvvv \
        --config=/githooks/phpcs.php_cs \
         2>&1

### Configuration file

```
<?php

declare(strict_types=1);

// INFO: the file-extension ".php_cs" is for PhpStorm, so that we can choose this file in the config :/

require_once __DIR__ . '/PhpCsExtras/fixer/InternalClassCasingFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/ModernizeStriposFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/ModernizeStrposFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/NoLeadingSlashInGlobalNamespaceFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/NoUselessCommentFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/NoUselessParenthesisFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/PhpdocNoIncorrectVarAnnotationFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/PhpdocVarAnnotationToAssertFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/RemoveEmptyDocBlockFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/RemoveSuperfluousDocBlockWhitespaceFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/SingleSpaceAfterStatementFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/SingleSpaceBeforeStatementFixer.php';
require_once __DIR__ . '/PhpCsExtras/fixer/UselessCodeAfterReturnFixer.php';

$InternalClassCasingFixer = new InternalClassCasingFixer();
$ModernizeStriposFixer = new ModernizeStriposFixer();
$ModernizeStrposFixer = new ModernizeStrposFixer();
$NoLeadingSlashInGlobalNamespaceFixer = new NoLeadingSlashInGlobalNamespaceFixer();
$NoUselessCommentFixer = new NoUselessCommentFixer();
$NoUselessParenthesisFixer = new NoUselessParenthesisFixer();
$PhpdocNoIncorrectVarAnnotationFixer = new PhpdocNoIncorrectVarAnnotationFixer();
$PhpdocVarAnnotationToAssertFixer = new PhpdocVarAnnotationToAssertFixer();
$RemoveEmptyDocBlockFixer = new RemoveEmptyDocBlockFixer();
$RemoveSuperfluousDocBlockWhitespaceFixer = new RemoveSuperfluousDocBlockWhitespaceFixer();
$SingleSpaceAfterStatementFixer = new SingleSpaceAfterStatementFixer();
$SingleSpaceBeforeStatementFixer = new SingleSpaceBeforeStatementFixer();
$UselessCodeAfterReturnFixer = new UselessCodeAfterReturnFixer();

$cacheFile = '/tmp/.php_cs_' . date('Y') . '_' . date('m') . '.cache';
$useCache = true;

return (new PhpCsFixer\Config())
    ->setUsingCache($useCache)
    ->setCacheFile($cacheFile)
    ->registerCustomFixers(
                            [
                                $InternalClassCasingFixer,
                                $ModernizeStriposFixer,
                                $ModernizeStrposFixer,
                                $NoLeadingSlashInGlobalNamespaceFixer,
                                $NoUselessCommentFixer,
                                $NoUselessParenthesisFixer,
                                $PhpdocNoIncorrectVarAnnotationFixer,
                                $PhpdocVarAnnotationToAssertFixer,
                                $RemoveEmptyDocBlockFixer,
                                $RemoveSuperfluousDocBlockWhitespaceFixer,
                                $SingleSpaceAfterStatementFixer,
                                $SingleSpaceBeforeStatementFixer,
                                $UselessCodeAfterReturnFixer,
                            ]
                        )
                        ->setRiskyAllowed(true)
                        ->setRules(
                            [
                                $InternalClassCasingFixer->getName()                  => true,
                                $ModernizeStriposFixer->getName()                     => true,
                                $ModernizeStrposFixer->getName()                      => true,
                                $NoLeadingSlashInGlobalNamespaceFixer->getName()      => true,
                                $NoUselessCommentFixer->getName()                     => true,
                                $NoUselessParenthesisFixer->getName()                 => true,
                                $PhpdocNoIncorrectVarAnnotationFixer->getName()       => true,
                                $PhpdocVarAnnotationToAssertFixer->getName()          => true,
                                $RemoveEmptyDocBlockFixer->getName()                  => true,
                                $RemoveSuperfluousDocBlockWhitespaceFixer->getName()  => true,
                                $SingleSpaceBeforeStatementFixer->getName()           => true,
                                $SingleSpaceAfterStatementFixer->getName()            => true,
                                $UselessCodeAfterReturnFixer->getName()               => true,
                                'align_multiline_comment'                                 => [
                                    'comment_type' => 'all_multiline',
                                ],
                                'array_indentation' => true,
                                'array_push'        => true,
                                'array_syntax'      => [
                                    'syntax' => 'short',
                                ],
                                'backtick_to_shell_exec' => true,
                                'binary_operator_spaces' => [
                                    'operators' => ['=>' => 'align_single_space_minimal'],
                                ],
                                'blank_line_after_namespace'       => true,
                                'blank_line_after_opening_tag'     => true,
                                'blank_line_before_statement'      => true,
                                'blank_line_between_import_groups' => true,
                                'braces' => [
                                    'allow_single_line_closure' => true
                                ],
                                'class_definition' => [
                                    'single_line' => true,
                                ],
                                'cast_spaces' => [
                                    'space' => 'none',
                                ],
                                'class_reference_name_casing' => true,
                                'class_attributes_separation' => [
                                    'elements' => [
                                        'const'    => 'one',
                                        'property' => 'one',
                                    ],
                                ],
                                'combine_consecutive_issets' => true,
                                'combine_consecutive_unsets' => false, // maybe better for readability, so keep it
                                'combine_nested_dirname'     => true,
                                'compact_nullable_typehint'  => true,
                                'concat_space'               => [
                                    'spacing' => 'one',
                                ],
                                'date_time_immutable'              => false, // do not change the logic of the code
                                'declare_equal_normalize'          => true,
                                'declare_strict_types'             => false, // do not change the logic of the code
                                'dir_constant'                     => true,
                                'elseif'                           => true,
                                'encoding'                         => true,
                                'ereg_to_preg'                     => true,
                                'error_suppression'                => false, // keep current error suppression for now
                                'escape_implicit_backslashes'      => false, // do not change the code
                                'explicit_indirect_variable'       => true,
                                'explicit_string_variable'         => true,
                                'final_internal_class'             => false, // todo: try to use it
                                'fopen_flag_order'                 => true,
                                'fopen_flags'                      => false, // todo: try to use it
                                'full_opening_tag'                 => true,
                                'fully_qualified_strict_types'     => false, // maybe better for readability, so keep it
                                'function_declaration'             => true,
                                'function_to_constant'             => true,
                                'function_typehint_space'          => true,
                                'general_phpdoc_annotation_remove' => [
                                    'annotations' => [
                                        'author',
                                        'package',
                                        'version',
                                    ],
                                ],
                                'get_class_to_class_keyword'  => false, // PHP 8.0
                                'heredoc_to_nowdoc'           => false, // IDEA: NOWDOC is simpler, so maybe use it?
                                'heredoc_indentation'         => true,
                                'clean_namespace'             => true,
                                'implode_call'                => true,
                                'include'                     => true,
                                'increment_style'             => false, // maybe better for readability, so keep it
                                'indentation_type'            => true,
                                'line_ending'                 => true,
                                'linebreak_after_opening_tag' => true,
                                'list_syntax'                 => [
                                    'syntax' => 'short',
                                ],
                                'logical_operators'          => true,
                                'lowercase_cast'             => true,
                                'lowercase_keywords'         => true,
                                'lowercase_static_reference' => true,
                                'magic_constant_casing'      => true,
                                'magic_method_casing'        => true,
                                'method_argument_space'      => [
                                    'on_multiline'                     => 'ignore',
                                    'keep_multiple_spaces_after_comma' => false,
                                ],
                                'method_chaining_indentation'            => false,
                                'modernize_types_casting'                => true,
                                'multiline_comment_opening_closing'      => false,
                                'multiline_whitespace_before_semicolons' => [
                                    'strategy' => 'no_multi_line',
                                ],
                                'native_constant_invocation'         => true,
                                'native_function_casing'             => true,
                                'native_function_invocation'         => false, // do not add "\" everywhere, maybe better for readability
                                'new_with_braces'                    => true,
                                'use_arrow_functions'                => true,
                                'no_alias_functions'                 => true,
                                'no_alias_language_construct_call'   => true,
                                'no_alternative_syntax'              => true,
                                'no_binary_string'                   => true,
                                'no_blank_lines_after_class_opening' => true,
                                'no_blank_lines_after_phpdoc'        => true,
                                'no_blank_lines_before_namespace'    => false, // we use "single_blank_line_before_namespace" instead
                                'no_break_comment'                   => true,
                                'no_closing_tag'                     => true,
                                'no_empty_comment'                   => true,
                                'no_empty_phpdoc'                    => true,
                                'no_empty_statement'                 => true,
                                'no_extra_blank_lines'               => true,
                                'no_homoglyph_names'                 => true,
                                'no_leading_import_slash'            => true,
                                'no_leading_namespace_whitespace'    => true,
                                'no_mixed_echo_print'                => [
                                    'use' => 'echo',
                                ],
                                'no_multiline_whitespace_around_double_arrow' => true,
                                'no_null_property_initialization'             => false,
                                'no_php4_constructor'                         => false, // PHP 8.0
                                'no_short_bool_cast'                          => true,
                                'echo_tag_syntax'                             => true,
                                'no_singleline_whitespace_before_semicolons'  => true,
                                'no_spaces_after_function_name'               => true,
                                'no_spaces_around_offset'                     => true,
                                'no_spaces_inside_parenthesis'                => true,
                                'no_space_around_double_colon'                => true,
                                'no_superfluous_elseif'                       => true,
                                'no_superfluous_phpdoc_tags'                  => [
                                    'allow_mixed'         => true,
                                    'allow_unused_params' => false,
                                ],
                                'no_trailing_comma_in_singleline'   => true,
                                'no_trailing_whitespace'            => true,
                                'no_trailing_whitespace_in_comment' => true,
                                'no_unneeded_control_parentheses'   => ['statements' => [
                                    'break',
                                    'clone',
                                    'continue',
                                    'echo_print',
                                    'negative_instanceof',
                                    'others',
                                    'return',
                                    'switch_case',
                                    'yield',
                                    'yield_from',
                                ]],
                                'no_unneeded_curly_braces'              => true,
                                'no_unneeded_final_method'              => true,
                                'no_unreachable_default_argument_value' => false, // do not change the logic of the code
                                'no_unset_cast'                         => true,
                                'no_unset_on_property'                  => true,
                                'no_unused_imports'                     => true,
                                'no_useless_else'                       => true,
                                'no_useless_concat_operator'            => [
                                    'juggle_simple_strings' => false,
                                ],
                                'no_useless_return'                   => true,
                                'no_useless_nullsafe_operator'        => true,
                                'no_whitespace_before_comma_in_array' => true,
                                'no_whitespace_in_blank_line'         => true,
                                'non_printable_character'             => true,
                                'normalize_index_brace'               => true,
                                'not_operator_with_space'             => false, // keep current code style for now
                                'not_operator_with_successor_space'   => false, // keep current code style for now
                                'object_operator_without_whitespace'  => true,
                                'ordered_class_elements'              => false, // maybe better for readability, so keep it
                                'ordered_imports'                     => true,
                                'phpdoc_add_missing_param_annotation' => [
                                    'only_untyped' => true,
                                ],
                                'phpdoc_align'                  => true,
                                'phpdoc_annotation_without_dot' => true,
                                'phpdoc_indent'                 => true,
                                'general_phpdoc_tag_rename'     => ['replacements' => ['inheritDocs' => 'inheritDoc']],
                                'phpdoc_inline_tag_normalizer'  => ['tags' => [
                                    'example',
                                    'id',
                                    'internal',
                                    'inheritdoc',
                                    'inheritdocs',
                                    'link',
                                    'source',
                                    'toc',
                                    'tutorial',
                                ]],
                                'phpdoc_tag_type' => [
                                    'tags' => [
                                        'api'        => 'annotation',
                                        'author'     => 'annotation',
                                        'copyright'  => 'annotation',
                                        'deprecated' => 'annotation',
                                        'example'    => 'annotation',
                                        'global'     => 'annotation',
                                        'inheritDoc' => 'inline',
                                        'internal'   => 'annotation',
                                        'license'    => 'annotation',
                                        'method'     => 'annotation',
                                        'package'    => 'annotation',
                                        'param'      => 'annotation',
                                        'property'   => 'annotation',
                                        'return'     => 'annotation',
                                        'see'        => 'annotation',
                                        'since'      => 'annotation',
                                        'throws'     => 'annotation',
                                        'todo'       => 'annotation',
                                        'uses'       => 'annotation',
                                        'var'        => 'annotation',
                                        'version'    => 'annotation',
                                    ],
                                ],
                                'phpdoc_order_by_value' => true,
                                'phpdoc_line_span'      => [
                                    'const'    => 'multi',
                                    'method'   => 'multi',
                                    'property' => 'multi',
                                ],
                                'phpdoc_no_access'             => true,
                                'phpdoc_no_alias_tag'          => true,
                                'phpdoc_no_empty_return'       => false,
                                'phpdoc_no_package'            => true,
                                'phpdoc_no_useless_inheritdoc' => true,
                                'phpdoc_order'                 => true,
                                'phpdoc_return_self_reference' => true,
                                'phpdoc_scalar'                => true,
                                'phpdoc_separation'            => [
                                    'groups' => [
                                        ['link', 'see', 'since'],
                                        ['author', 'copyright', 'license'],
                                        ['category', 'package', 'subpackage'],
                                        ['deprecated', 'internal', 'todo'],
                                        ['db-type', 'db-type-simple', 'db-size', 'db-decimal', 'db-default'],
                                        ['ValueObjectType', 'ValueObjectDefault', 'phpstan-var', 'skip-php-doc-auto-update'],
                                        ['property', 'property-read', 'property-write'],
                                        ['phpstan-property', 'phpstan-property-read', 'phpstan-property-write'],
                                        ['psalm-property', 'psalm-property-read', 'psalm-property-write'],
                                        ['template-covariant', 'template', 'phpstan-template', 'psalm-template', 'param-out', 'phpstan-param', 'phpstan-param-out', 'psalm-param', 'psalm-param-out', 'phpstan-return', 'psalm-return', 'extends', 'phpstan-extends', 'template-extends', 'implements', 'phpstan-implements', 'template-implements'],
                                    ],
                                ],
                                'phpdoc_single_line_var_spacing'                => true,
                                'phpdoc_summary'                                => false,
                                'phpdoc_to_comment'                             => true,
                                'phpdoc_to_return_type'                         => false, // maybe not working as expected (because of wrong phpdoc comments)
                                'phpdoc_trim'                                   => true,
                                'phpdoc_trim_consecutive_blank_line_separation' => true,
                                'phpdoc_types'                                  => true,
                                'phpdoc_types_order'                            => [
                                    'null_adjustment' => 'always_first',
                                    'sort_algorithm'  => 'alpha',
                                ],
                                'phpdoc_var_annotation_correct_order' => true,
                                'phpdoc_var_without_name'             => true,
                                'pow_to_exponentiation'               => true,
                                'protected_to_private'                => true,
                                'return_assignment'                   => false, // not working as expected
                                'return_type_declaration'             => true,
                                'self_accessor'                       => true,
                                'self_static_accessor'                => true,
                                'semicolon_after_instruction'         => true,
                                'set_type_to_cast'                    => true,
                                'short_scalar_cast'                   => true,
                                'simplified_null_return'              => false, // maybe better for readability, so keep it
                                'single_blank_line_at_eof'            => true,
                                'single_blank_line_before_namespace'  => true,
                                'single_class_element_per_statement'  => true,
                                'single_import_per_statement'         => true,
                                'single_line_after_imports'           => true,
                                'single_line_comment_spacing'         => false, // use it in the future
                                'single_line_comment_style'           => [
                                    'comment_types' => ['hash'],
                                ],
                                'single_line_throw'      => true,
                                'single_quote'           => false, // keep quotes for now
                                'space_after_semicolon'  => true,
                                'standardize_increment'  => false, // maybe better for readability, so keep it
                                'standardize_not_equals' => true,
                                'static_lambda'          => true,
                                // 'strict_comparison'   => true, // do not change the logic of the code
                                // 'strict_param'        => true, // do not change the logic of the code
                                'string_length_to_empty'                   => true,
                                'string_line_ending'                       => true,
                                'switch_case_semicolon_to_colon'           => true,
                                'switch_case_space'                        => true,
                                'switch_continue_to_break'                 => true,
                                'ternary_operator_spaces'                  => true,
                                'ternary_to_null_coalescing'               => true,
                                'assign_null_coalescing_to_coalesce_equal' => true,
                                'trailing_comma_in_multiline'              => true,
                                'trim_array_spaces'                        => true,
                                'unary_operator_spaces'                    => true,
                                'visibility_required'                      => [
                                    'elements' => [
                                        'const',
                                        'method',
                                        'property',
                                    ],
                                ],
                                'void_return'                     => false, // todo: try to use it
                                'whitespace_after_comma_in_array' => true,
                                'yoda_style'                      => [
                                    'equal'            => false,
                                    'identical'        => false,
                                    'less_and_greater' => false,
                                ],
                            ]
                        )
    ->setFinder(
                            PhpCsFixer\Finder::create()
                                ->in(INSTALL_DIR)
                                ->notPath('#.*/vendor/.*#')
                                ->notPath('#.*/tmp/.*#')
                                ->name('*.php')
                                ->name('*.inc')
                                ->ignoreDotFiles(true)
                                ->ignoreVCS(true)
                        );

```

### Code snippet that reproduces the problem

  ```php
  <?php
  /**
   * @return 'NULL'|null|false|int
   */
  function foo() {
   // ...
   }
  ```

### PR DESCRIPTION
**Pull request by mvorisek**
fix #7101

### CODE REVIEW

File: tests/Fixer/Phpdoc/PhpdocTypesFixerTest.php [hunk 1]
```diff
@@ -259,7 +259,25 @@ public static function provideFixCases(): iterable
...
+                    */',
+        ];
+
+        yield 'union with "NULL"' => [
+            '<?php /**
+                    * @return null|"NULL"|false
+                    */',
+            '<?php /**
+                    * @return NULL|"NULL"|false
                     */',
```
Comment (mvorisek) [inline tests/Fixer/Phpdoc/PhpdocTypesFixerTest.php:281]: needs `@method` test, 2nd callable type must not be lowercased

**Comment by Wirone** [comment]
@mvorisek is it ready for review? I don't review drafts by definition 😉.

**Comment by mvorisek** [comment]
PR is done. The fixing is done on correctly parsed type expressions so we can be sure it is never done inside const strings.

File: tests/DocBlock/TypeExpressionTest.php [hunk 2]
```diff
@@ -591,12 +591,20 @@ public static function provideAllowsNullCases(): iterable
      */
     public function testSortTypes(string $typesExpression, string $expectResult): void
     {
+        $sortCaseFx = static function (TypeExpression $a, TypeExpression $b): int {
+            return strcasecmp($a->toString(), $b->toString());
+        };
+        $sortCrc32Fx = static function (TypeExpression $a, TypeExpression $b): int {
+            return strcasecmp($a->toString(), $b->toString());
+        };
```
Comment (Wirone) [inline tests/DocBlock/TypeExpressionTest.php:599]:
```suggestion
```

`$sortCrc32Fx` is the same as `$sortCaseFx`, is it needed here?

Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php]: typo in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/commit/f08feb3a585cb232852b974d53d2d5f11659edb9, good catch!

**Comment by Wirone** [comment]
@mvorisek I'll try to look at it in the evening.

**Review by kubawerlos — APPROVED** [review]

**Review by Wirone — APPROVED** [review]

**Comment by Wirone** [comment]
Thank you @mvorisek 🍻!

PS. it was [long afternoon](#issuecomment-1665948956) 😅.

**Pull request merged**